compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp

```diff
@@ -0,0 +1,608 @@
+/// fusion root may be null. If the fusion root is null, the reduction operation
+/// should be used as fusion root instead.
+static std::tuple<Value, Value, Value> reductionBlockDistributionHandles(
+    ImplicitLocOpBuilder &b, Value variantH, bool hasLeadingEltwise,
+    bool hasTrailingEltwise) {
+  Value originalGenericH =
+      b.create<MatchOp>(variantH, linalg::GenericOp::getOperationName());
+  auto op = b.create<SplitHandlesOp>(
+      originalGenericH,
+      /*numResultHandles=*/1 + hasLeadingEltwise + hasTrailingEltwise);
+  return std::make_tuple(hasLeadingEltwise ? op.getResults().front() : Value(),
+                         op.getResults().drop_front(hasLeadingEltwise).front(),
+                         hasTrailingEltwise ? op.getResults().back() : Value());
+}
+
+// TODO: generalize and automate over and over.
+// TODO: significantly shrink this down.
+static void buildReductionCudaStrategy(ImplicitLocOpBuilder &b, Value variantH,
+                                       const GPUReductionStrategyInfos &infos) {
+  // Step 0. Match the ops.
+  Value originalFillH =
+      b.create<MatchOp>(variantH, linalg::FillOp::getOperationName());
+  auto [leadingH, reductionH, fusionRootH] = reductionBlockDistributionHandles(
+      b, variantH, infos.hasLeadingEltwise, infos.hasTrailingEltwise);
+
+  // Step 1: Distribute to blocks using the current IREE lowering config.
+  variantH = buildReductionStrategyBlockDistributionPart(
+      b, variantH, originalFillH, reductionH, fusionRootH,
+      infos.workgroupTileSizes, infos.hasLeadingEltwise,
+      infos.hasTrailingEltwise);
+
+  // Step 2. Second level of tiling + fusion parallelizes to threads.
+  variantH = buildReductionStrategyThreadDistributionPart(
+      b, variantH, infos.fillSecondTileSizes, infos.genericSecondTileSizes,
+      infos.hasLeadingEltwise, infos.hasTrailingEltwise);
+
+  // Step 3. Rank-reduce and vectorize.
+  // TODO: assumes a single func::FuncOp to transform, may need hardening.
+  Value funcH = b.create<MatchOp>(variantH, func::FuncOp::getOperationName());
+  funcH = buildVectorizeStrategy(b, funcH);
+
+  // Step 4. Bufferize and drop HAL decriptor from memref ops.
+  variantH = b.create<IREEBufferizeOp>(variantH, /*targetGpu=*/true);
+  Value memrefFunc =
+      b.create<MatchOp>(variantH, func::FuncOp::getOperationName());
+  b.create<IREEEraseHALDescriptorTypeFromMemRefOp>(memrefFunc);
+
+  // Step 5. Post-bufferization mapping to blocks and threads.
+  // Need to match again since bufferize invalidated all handles.
+  // TODO: assumes a single func::FuncOp to transform, may ned hardening.
+  funcH = b.create<MatchOp>(variantH, func::FuncOp::getOperationName());
+  funcH = mapToBlockAndThreads(b, funcH, infos.workgroupSize);
+
+  // Step 6. Post-bufferization vector distribution with rank-reduction.
+  distributeVectors(b, funcH);
+}
+
+static constexpr unsigned cudaWarpSize = 32;
```

Comment:
style nit: naming with a leading "k" followed by mixed case.

```suggestion
static constexpr unsigned kCudaWarpSize = 32;
```